File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    
    // let L = 0, R = 1;
    // let max = 0;

    // while( R < prices.length ) {
    //     if ( prices[R] - prices[L] < 0 ) {
    //         // console.log(prices[R] - prices[L])
    //         L = R;
    //     } else {
    //         if ( prices[R] - prices[L] > max ) {
    //             // console.log(prices[R] - prices[L])
    //             max = prices[R] - prices[L];
    //         }
    //     }
    //     R++;
    // }
    // return max
    

//    let left = 0, right = 1, max = 0;
//    while ( right < prices.length ) {
//         if ( prices[right] - prices[left] < 0 ) {
//             left = right;
//         } else {
//             max = Math.max(max, prices[right] - prices[left])
//         }
//      right++;
//    }
//    return max;





    let left=0, right=1, max=0;
    while ( right < prices.length ) {
        if ( prices[left] > prices[right] ) {
            left = right;
        }
        else {
            max = Math.max(max, prices[right]-prices[left]);
            right++;
        }
    }

    return max;








};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int[] suffix = new int[n];
        suffix[n-1] = prices[n-1];
        for ( int i=n-2; i>=0; i-- ) {
            suffix[i] = Math.max(suffix[i+1], prices[i]);
        }
        // [7,6,6,6,6,4]
        int maxVal = 0;
        for ( int i=0; i<n; i++ ) {
            maxVal = Math.max(maxVal, suffix[i] - prices[i]);
        }
        return maxVal;
    }
}